packages/mui-material/src/styles/stringifyTheme.ts

```diff
@@ -0,0 +1,62 @@
+/* eslint-disable import/prefer-default-export */
+import { isPlainObject } from '@mui/utils/deepmerge';
+
+function isSerializable(val: any) {
+  return (
+    isPlainObject(val) ||
+    typeof val === 'undefined' ||
+    typeof val === 'string' ||
+    typeof val === 'boolean' ||
+    typeof val === 'number' ||
+    Array.isArray(val)
+  );
+}
+
+/**
+ * `baseTheme` usually comes from `createTheme` or `extendTheme`.
+ *
+ * This function is intended to be used with zero-runtime CSS-in-JS like Pigment CSS
+ * For example, in a Next.js project:
+ *
+ * ```js
+ * // next.config.js
+ * const { extendTheme } = require('@mui/material/styles');
+ *
+ * const theme = extendTheme();
+ * // `.toRuntimeSource` is Pigment CSS specific to create a theme that is available at runtime.
+ * theme.toRuntimeSource = stringifyTheme;
+ *
+ * module.exports = withPigment({
+ *  theme,
+ * });
+ * ```
+ */
+export function stringifyTheme(baseTheme: Record<string, any> = {}) {
+  const serializableTheme: Record<string, any> = { ...baseTheme };
+
+  function serializeTheme(object: Record<string, any>) {
+    const array = Object.entries(object);
+    // eslint-disable-next-line no-plusplus
+    for (let index = 0; index < array.length; index++) {
+      const [key, value] = array[index];
+      if (!isSerializable(value) || key.startsWith('unstable_')) {
+        delete object[key];
+      } else if (isPlainObject(value)) {
+        object[key] = { ...value };
+        serializeTheme(object[key]);
+      }
+    }
+  }
```

Comment:
I think we need to remove all the CSS variables values, colors, typography, etc., so it breaks at runtime if developers try to access them.

Reply:
But how would you make it work for components that are using `useTheme()`?

Reply:
> React.cache is still experimental

https://nextjs.org/docs/app/building-your-application/caching#react-cache-function doesn't make it so clear.

> But how would you make it work for components that are using useTheme() at runtime, for example, Material UI Tooltip?

For the client bundle, you have the whole theme available, it propagates in the application with then context API. 
For the server bundle, you have a full duplicate of the theme object , it's made available ideally with the React.cache API, but with a single theme file also works. 

The job of the plugin is to duplicate the theme file, not even to serialized it.

Reply:
cc @DiegoAndai